File: pt5/1182.shortest-distance-to-target-color.go
package main

import "sort"

func shortestDistanceColor(colors []int, queries [][]int) []int {
	positions := map[int][]int{1: []int{}, 2: []int{}, 3: []int{}}
	for index, color := range colors {
		positions[color] = append(positions[color], index)
	}

	answer := make([]int, len(queries))
	for i, query := range queries {
		index, color := query[0], query[1]
		colorPositions := positions[color]
		if len(colorPositions) == 0 {
			answer[i] = -1
			continue
		}

		insertAt := sort.SearchInts(colorPositions, index)
		best := 1 << 30
		if insertAt < len(colorPositions) && colorPositions[insertAt]-index < best {
			best = colorPositions[insertAt] - index
		}
		if insertAt > 0 && index-colorPositions[insertAt-1] < best {
			best = index - colorPositions[insertAt-1]
		}
		answer[i] = best
	}
	return answer
}

File: pt5/1182.shortest-distance-to-target-color.java
import java.util.*;

class Solution {
    @SuppressWarnings("unchecked")
    public List<Integer> shortestDistanceColor(int[] colors, int[][] queries) {
        List<Integer>[] positions = new ArrayList[4];
        for (int color = 1; color <= 3; color++) {
            positions[color] = new ArrayList<>();
        }
        for (int i = 0; i < colors.length; i++) {
            positions[colors[i]].add(i);
        }

        List<Integer> answer = new ArrayList<>(queries.length);
        for (int[] query : queries) {
            int index = query[0];
            int color = query[1];
            List<Integer> colorPositions = positions[color];
            if (colorPositions.isEmpty()) {
                answer.add(-1);
                continue;
            }

            int insertAt = Collections.binarySearch(colorPositions, index);
            if (insertAt < 0) {
                insertAt = -insertAt - 1;
            }

            int best = Integer.MAX_VALUE;
            if (insertAt < colorPositions.size()) {
                best = Math.min(best, colorPositions.get(insertAt) - index);
            }
            if (insertAt > 0) {
                best = Math.min(best, index - colorPositions.get(insertAt - 1));
            }
            answer.add(best);
        }

        return answer;
    }
}
